synapse/storage/events.py

```diff
@@ -596,10 +604,77 @@ def _get_events(txn, batch):
             )
 
         for chunk in batch_iter(event_ids, 100):
-            yield self.runInteraction("_get_events_which_are_prevs", _get_events, chunk)
+            yield self.runInteraction(
+                "_get_events_which_are_prevs",
+                _get_events_which_are_prevs_txn,
+                chunk,
+            )
 
         defer.returnValue(results)
 
+    @defer.inlineCallbacks
+    def _get_prevs_before_rejected(self, event_ids):
+        """Get soft-failed ancestors to remove from the extremities.
+
+        Given a set of events recursively find all prev events that have
```

Comment:
> So this starts with the events we're about to persist and checks their prev events

ok, but the events passed into this method are the prev events of the event we are about to persist, and the first ones we check. So it's not true to say "Given a set of events recursively find all prev events that have been rejected", rather:
```suggestion
        Given a set of events recursively find all those that have
```

Reply:
Or rather:

```
Given a set of events, find all those that have been soft-failed or rejected. Returns those soft failed/rejected events
and their prev events (whether soft-failed/rejected or not), and recurses up the prev-event graph until it finds no more soft-failed/rejected events.
```